include/dlaf/communication/kernels/reduce.h

```diff
@@ -37,127 +34,194 @@ namespace comm {
+  // (CPU/cCPU) --> MPI --> (CPU/cCPU)
   //
-  // CHECKPOINT assures that the copyBack has finished, which is useful for the next step.
+  // where: cCPU = contiguous CPU
 
-  auto cont_buf = dataflow(ex_copy, unwrapping(makeItContiguous_o), tile_cpu);
-  cont_buf =
-      dataflow(ex, unwrapping(reduceRecvInPlace_o), std::move(pcomm), reduce_op, std::move(cont_buf));
-  auto checkpoint = dataflow(ex_copy, unwrapExtendTiles(copyBack_o), std::move(cont_buf), tile_cpu);
+  using namespace pika::execution::experimental;
 
+  internal::senderReduceRecvInPlace<T>(ex, std::move(pcomm), reduce_op, std::move(tile)) |
+      start_detached();
+}
+
+/// Given a GPU tile, perform MPI_Reduce in-place
+template <class T, Device D>
+void scheduleReduceRecvInPlace(const comm::Executor& ex,
+                               pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
+                               MPI_Op reduce_op, pika::future<matrix::Tile<T, D>> tile) {
   // Note:
   //
-  // In case the original tile was on CPU, the previous step already managed to copy the result in place
-  // and nothing else has to be done.
-  // Otherwise, if it was on GPU, a temporary tile on CPU has been used, so this step handles the needed
-  // copy from CPU to GPU.
+  // GPU --> Duplicate --> (cCPU --> MPI --> cCPU) --> copy --> GPU
   //
-  // TILE_CPU => no-op
-  // TILE_GPU => copy back from CPU to GPU
+  // where: cCPU = contiguous CPU
 
-  copyIfNeeded(tile_cpu, tile_orig, std::move(checkpoint));
+  using namespace pika::execution::experimental;
+  using dlaf::internal::transform;
+
+  std::move(tile) |
+      let_value(pika::unwrapping([ex, reduce_op, pcomm = std::move(pcomm)](auto& tile_gpu) mutable {
+        using dlaf::internal::Policy;
+        using dlaf::matrix::internal::CopyBackend;
+
+        // GPU -> cCPU
+        auto tile_cpu = transform(
+            Policy<CopyBackend<D, Device::CPU>::value>(pika::threads::thread_priority::high),
+            [](const matrix::Tile<const T, Device::GPU>& tile_gpu, auto... args) mutable {
+              return dlaf::matrix::Duplicate<Device::CPU>{}(tile_gpu, args...);
+            },
+            just(std::cref(tile_gpu)));
+
+        // cCPU -> MPI -> cCPU
+        auto tile_reduced =
+            internal::senderReduceRecvInPlace<T>(ex, std::move(pcomm), reduce_op, std::move(tile_cpu));
+
+        // cCPU -> GPU
+        namespace arg = std::placeholders;
+        return transform(Policy<CopyBackend<Device::CPU, D>::value>(
+                             pika::threads::thread_priority::high),
+                         std::bind(matrix::internal::copy_o, arg::_1, std::cref(tile_gpu), arg::_2),
+                         std::move(tile_reduced));
+      })) |
+      start_detached();
 }
 
-template <class T, Device D, template <class> class Future>
+// TODO scheduleReduceSend with future will require to move the actual value, not the cref
```

Comment:
Can you expand on this comment? I'm not sure I get it.

Reply:
Thanks, makes sense! I suppose this can be refactored afterwards as long as the TODO is not actually a blocker for the other functionality here?

Reply:
Yes, exactly. Do you want to keep track of it with an issue? I mean, not just target this specific function, but in general about "allowing a function that requires a readonly tile via a sender, to accept both shared_future and future".

Reply:
I think it's covered by https://github.com/eth-cscs/DLA-Future/issues/465 since that will require quite significant refactoring of these functions anyway. Do you agree?